README.md
# React Native Coachmark

![NPM Version](https://badge.fury.io/js/react-native-coachmark.svg) [![CircleCI](https://circleci.com/gh/traveloka/react-native-coachmark/tree/master.svg?style=shield&circle-token=ab2fa2d1fdea74636ee3132e16c82003cdc95383)](https://circleci.com/gh/traveloka/react-native-coachmark) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)

![preview](docs/demo.gif)

## Installation

yarn

```
yarn add react-native-coachmark
```

npm

```
npm i react-native-coachmark --save
```

## Getting Started

### Basic Usage

```javascript
import React from 'react';
import { Coachmark } from 'react-native-coachmark';

export default function AwesomeScreen() {
  return (
    <Page>
      <Header>
        <Coachmark autoShow message="Click here to save your page!">
          <ButtonBookmark>
        </Coachmark>
      </Header>
    </Page>
  );
}
```

## Example

[![preview](https://i.ibb.co/n3VpkY4/Screen-Shot-2018-12-04-at-13-53-26.png)](https://snack.expo.io/@jekiwijaya/react-native-coachmark)

[View on Expo!](https://snack.expo.io/@jekiwijaya/react-native-coachmark)

## Documentation

### - Coachmark

#### Importing

```javascript
var Coachmark = require('react-native-coachmark').Coachmark; // ES5

import { Coachmark } from 'react-native-coachmark'; // ES6
```

#### Props

| Property         | Type     | Default Value | Description                                    |
| ---------------- | -------- | ------------- | ---------------------------------------------- |
| message          | string   | none          | required                                       |
| autoShow         | boolean  | none          | to show the coachmark when mounting            |
| showArrow        | boolean  | true          | to show the arrow in the coachmark             |
| coachmarkContent | function | null          | to show a custom content for coachmark         |
| onShow           | function | none          | will be called when coachmark is showing       |
| onHide           | function | none          | will be called when coachmark is hidden        |
| isAnchorReady    | boolean  | true          | a value to force coachmark not being shown yet |

#### Methods

| Methods                      | Description                               |
| ---------------------------- | ----------------------------------------- |
| `show() => Promise`          | a function to trigger show the coachmark  |
| `coachmarkContent() => null` | a function to render content in coachmark |

### Roadmap

- [ ] Auto load and save in AsyncStorage
- [x] Show coachmark only when in view port
- [ ] Custom render arrow and content

## Contributing

We'd to have your helping hand on this package! Feel free to PR's, add issues or give feedback!

## Credits

Written by [Jacky Wijaya](https://www.linkedin.com/in/jacky-wijaya-125b90b6/) (jekiwijaya) at Traveloka.


CHANGELOG.md
# [0.1.0-alpha.1](https://github.com/traveloka/react-native-coachmark/compare/v0.1.0-alpha.0...v0.1.0-alpha.1) (2021-03-05)



# [0.1.0-alpha.0](https://github.com/traveloka/react-native-coachmark/compare/v0.0.4...v0.1.0-alpha.0) (2021-03-02)


### Bug Fixes

* circleci image ([#18](https://github.com/traveloka/react-native-coachmark/issues/18)) ([76f4c42](https://github.com/traveloka/react-native-coachmark/commit/76f4c426d269551a2caa5f659679e0c3882bf1e4))



## [0.0.4](https://github.com/traveloka/react-native-coachmark/compare/v0.0.3...v0.0.4) (2019-09-09)



## [0.0.3](https://github.com/traveloka/react-native-coachmark/compare/v0.0.2...v0.0.3) (2019-01-15)



## [0.0.2](https://github.com/traveloka/react-native-coachmark/compare/v0.0.1...v0.0.2) (2019-01-14)



## 0.0.1 (2018-12-03)





src/types.ts
import React from 'react';
import { StyleProp, ViewStyle } from 'react-native';

export interface CoachmarkProps {
  message?: string;
  autoShow?: boolean;
  onHide?: () => void;
  onShow?: () => void;
  isAnchorReady?: boolean;
  renderArrow?: CoachmarkViewProps['renderArrow'];
  accessibilityLabel?: string;
  testID?: string;
  contentContainerStyle?: StyleProp<ViewStyle>;
}

export enum CoachmarkPosition {
  TOP = 'top',
  BOTTOM = 'bottom',
}

export interface CoachmarkArrowProps {
  position?: CoachmarkPosition;
  x: number;
}

export interface CoachmarkContentProps {
  message: string;
  buttonText?: string;
}

export type CoachmarkViewProps = {
  renderArrow: ({
    x,
    position,
  }: {
    x: number;
    position?: CoachmarkPosition;
  }) => React.ReactElement<CoachmarkArrowProps>;
} & CoachmarkContentProps &
  CoachmarkArrowProps;


src/components/CoachmarkView.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CoachmarkContent from './CoachmarkContent';
import CoachmarkArrow, { K_POSITION_TOP, K_POSITION_BOTTOM } from './CoachmarkArrow';

export default class CoachmarkView extends Component {
  static propTypes = {
    position: PropTypes.oneOf([K_POSITION_TOP, K_POSITION_BOTTOM]),
    renderArrow: PropTypes.func,
  };

  static defaultProps = {
    position: K_POSITION_TOP,
    renderArrow: ({ x, position }) => <CoachmarkArrow x={x} position={position} />,
  };

  renderCoachmarkContent() {
    return <CoachmarkContent message={this.props.message} />;
  }

  renderCoachmarkArrow() {
    const { renderArrow, ...rest } = this.props;

    return renderArrow(rest);
  }

  render() {
    return this.props.position === K_POSITION_TOP ? (
      <React.Fragment>
        {this.renderCoachmarkArrow()}
        {this.renderCoachmarkContent()}
      </React.Fragment>
    ) : (
      <React.Fragment>
        {this.renderCoachmarkContent()}
        {this.renderCoachmarkArrow()}
      </React.Fragment>
    );
  }
}


src/components/CoachmarkArrow.js
import React, { Component } from 'react';
import { View, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';

export const K_POSITION_TOP = 'top';
export const K_POSITION_BOTTOM = 'bottom';

export default class CoachmarkArrow extends Component {
  static propTypes = {
    position: PropTypes.oneOf([K_POSITION_TOP, K_POSITION_BOTTOM]),
  };

  static defaultProps = {
    position: K_POSITION_TOP,
  };

  render() {
    const styles = getStyles(this.props.position);
    return <View style={[styles.arrow, { marginLeft: this.props.x - 10 }]} />;
  }
}

const getStyles = position =>
  StyleSheet.create({
    arrow: {
      width: 0,
      height: 0,
      borderRightWidth: 10,
      borderLeftWidth: 10,
      borderTopColor: 'transparent',
      borderRightColor: 'transparent',
      borderLeftColor: 'transparent',
      ...(position === K_POSITION_TOP
        ? { borderBottomColor: '#FFF', borderBottomWidth: 10 }
        : { borderTopColor: '#FFF', borderTopWidth: 10 }),
      ...(position === K_POSITION_TOP ? { marginTop: 12 } : { marginBottom: 12 }),
    },
  });


src/components/withStopableScrollView.js
import React from 'react';
import PropTypes from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';

export default function(WrappedComponent) {
  const StopableScroll = class extends React.Component {
    static propTypes = {
      onScroll: PropTypes.func,
    };

    static defaultProps = {
      onScroll: () => {}, // eslint-disable-line no-empty-function
    };

    constructor(props) {
      super(props);
      this.yOffset = 0;
      this.scrollView = React.createRef();
    }

    stop = () => {
      (this.props.scrollViewRef || this.scrollView).current.scrollTo({ x: 0, y: this.yOffset, animated: false });
    };

    _handleOnScroll = e => {
      this.yOffset = e.nativeEvent.contentOffset.y;
      this.props.onScroll(e);
    };

    render() {
      return (
        <WrappedComponent
          {...this.props}
          ref={this.props.scrollViewRef || this.scrollView}
          onScroll={this._handleOnScroll}
          scrollEventThrottle={16}
        />
      );
    }
  };
  return hoistNonReactStatics(StopableScroll, WrappedComponent);
}


src/components/CoachmarkContent.js
import React, { Component } from 'react';
import { Text, View, StyleSheet } from 'react-native';

export default class CoachmarkContent extends Component {
  render() {
    return (
      <View style={styles.container}>
        <View style={styles.message}>
          <Text style={styles.messageText}>{this.props.message}</Text>
        </View>
        <View style={styles.button}>
          <Text style={styles.buttonText}>OK</Text>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    borderRadius: 6,
    marginHorizontal: 16,
    backgroundColor: '#FFF',
    flexDirection: 'row',
    overflow: 'hidden',
  },
  message: {
    paddingHorizontal: 8,
    paddingVertical: 16,
    flex: 1,
  },
  messageText: {
    fontSize: 14,
    lineHeight: 22,
    letterSpacing: -0.15,
  },
  button: {
    width: 48,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'rgb(246,246,246)',
  },
  buttonText: {
    fontSize: 14,
    lineHeight: 22,
    color: 'rgb(7, 112, 205)',
    fontWeight: 'bold',
  },
});
